feat: add HTTP firmware update

sfoster1 opened the pull request:
This PR is an extremely large refactor of all three of the primary python packages that run on a robot to enable firmware update of connected microcontrollers via HTTP, and to reduce the amount of canbus traffic that only exists to poll what instruments are attached. 

It does all of this by adding a new first-class concept to the hardware management of the robot: the subsystem. The subsystem is an enum like the axes, and it represents a self-contained program running on a processor logically or physically distinct from the robot server (so you can't talk to it via normal linux commands, basically). In practice, that just means the microcontrollers on the flex; there are pretty easy extensions to have this also handle the motor controller board on the OT-2, and in fact I added an entry to the enum for just that purpose.

The subsystems have a firmware version, and an ok status, and can be updated. Internally, a subsystem being "not ok" really means "stuck in the bootloader"; this is a reasonable thing to say of a subsystem, because it's a program. It wouldn't be ok to say of an instrument, because in that case you wouldn't be able to know any of the information we need about a pipette: what its serial is, what its model is, things like that.

Subsystems are handled in hardware by a new object that handles their lifecycles and update cycles and manages discovering them, including asynchronously, so that interfaces above this don't need to actively poll.

We need a way for clients to figure out that they need to interact with the subsystems API, and so now when instruments that need a firmware update - whether that's because they're out of date, or because they're not booting properly - we return a `BadPipette` or `BadGripper` from `GET /instruments`, distinguishable by the presence of `ok: false` in the response (`ok: true` means the response is as before). This bad response also gives a route to the appropriate subsystem handler (see below).

Finally, this PR makes some new routes for clients to update subsystems - mostly instruments. These routes combine a way to view ongoing updates and a way to view historical updates. Ongoing updates are findable by subsystem, but when they complete they disappear; historical updates are viewable by key, and stick around while the robot is booted.

The client workflow for a client starting an update here would be:
- See that an update is needed by checking `GET /subsystems/status`
- Start an update by `POST /subsystems/updates/:subsystem`, and get an ID
- Check on the update by `GET /subsystems/updates/all/:id`

The client workflow for a client seeing that somebody else started an update would be:
- See that an update is going by checking `GET /subsystems/updates/current`
- Get its status by checking `GET /subsystems/updates/all/:id`

If two clients race to start an update, that's fine because the loser gets redirected to the ongoing update.

## Reviewing 

There is a lot here, and it's all intermingled. I separated it into commits that can be read one by one, but keep in mind that it wasn't developed that way; those commits probably don't stand alone in every detail, so if you see something that seems weird just check if it's in the full diff real quick.

This is not yet mergeable due to the testing and the last TODOs, but it really is ready in all respects for review.

## Wontdos

- I do not think I'm going to do tavern tests for this, for two reasons. First, the frontend isn't there yet; that means we may find things we need to change. Second, this PR's been under dev for a while and we really need to get it merged. Adding tavern tests is going to require a pretty robust sim/mock update backend for the sim server and it would probably add several days here.

## Testing

- [x] Run on an OT3 that has everything up to date and attach and remove pipettes and grippers that are up to date, checking that they are detected
- [x] Run on an OT3 that needs updates and check that it gets updated, and that the server is ready while those updates are happening (requires a TODO)
- [x] Run on an OT3 and connect pipettes and grippers that are out of date and update them via the HTTP interface (requires a TODO)
- [x] Run on an OT-3 and connect grippers and pipettes that are stuck in the bootloader and make sure they can be updated via the HTTP interface

## TODO
- [x] Change the initial update of connected devices to go through the robot server rather than the hardware controller so that they (a) don't block execution of other routes making you think the robot isn't on yet (b) or make it take so long to start systemd thinks it crashed (c) are visible in http
- [x] Write a nice little script to drive the http updates for testing
- ~[ ] Tavern integration tests~
     - [x] Ha ha it turns out the tavern tests don't and never have actually made ot3 hardware control simulators so these changes make tavern fail for all ot3 test runs so either make the simulator do that or add this stuff to ot2 also (done in https://github.com/Opentrons/opentrons/pull/12754 )
- [x] Add alternate response models from ~/pipettes~ (not used on ot3, don't need) and /instruments and so on that contain jsonapi links to subsystems in the case that a pipette is physically connected but either subsystem-status-bad or out of date

sfoster1 committed:
refactor(api): ThreadManager handles async gens

Async generator functions are different from normal async functions, in the same
way that generator functions are different from normal functions. When called,
they return a special async generator object. This is a problem in
ThreadManager land, because those async generator objects need to be
interacted with via `await anext(gen)` in the same loop that created
them.

So, this commit adds a new case in thread manager for "what if you call
an async generator function in the hardware controller". In that case,
what happens is
- We return a proxy async generator function, which is defined in (and
therefore runs in) the calling loop
- We generate a future that is a coroutine running in the hardware
controller loop that captures a queue, and pulls items out of the async
generator in that loop and puts them in a queue
- Our proxy pulls stuff out of the queue and yields this

This is of course extra frustrating because all queue interaction must
take place in the same loop, which must be the loop the queue was
created in. That means that for the queue that moves generator output
from the hardware thread/loop to the calling thread/loop, it has to be
- Created in the hardware thread/loop (via run_coroutine_threadsafe and
a wrapper)
- Inserted into in the hardware thread/loop (fine, this happens in the
coroutine that drains the generator)
- Gotten from in the hardware thread/loop, so every await queue.get()
needs to happen in a run_coroutine_threadsafe()

It all works! But it's pretty annoying.

sfoster1 committed:
fix(opentrons_hardware): errors during fw update

We weren't ever checking any exceptions that got raised during the
run-update task. We can do that by getting the result of the future
afterward (and throwing it away - the only thing we care about is that
result() will raise any exception that happened in the task).

sfoster1 committed:
refactor(robot-server): global IDNotFound error

We have these scattered hither and yon, let's get it in one place and
start reusing it.

sfoster1 committed:
refactor(api,robot-server): make SubSystem exposed

We had this type called OT3SubSystem, which was a representation of the
subsystems (programmable devices, really) that could be bolted on to the
machine. It was reasonably useful but never used to its full extent.

Now, let's use it. First, we're going to want it to work on the OT-2 as
well (just for completeness) so let's remove the OT-3 part of the name
and add a way to get to it from a mount; we'll leave a temporary alias
for other code.

As an example of how we'll use it, we can make a new set of models in
the server under subsystems/.

sfoster1 committed:
feat(robot-server): expose instrument subsystems

Add a new element "subsystem" to the instrument responses that lets
clients link an instrument to a subsystem. The subsystem is the
representation of the programmable microcontroller on the machine, and
is the part you can update. This is an important distinction to make
because if a subsystem isn't working properly, you generally can't
identify it as an instrument; but you can always identify it as a
subsystem. The subsystem retrieved here will be useful later in some new
routes.

Note: this commit definitely won't pass tests and mypy or like work on
its own, because this is a big interconnected PR.

sfoster1 committed:
refactor(api): remove fw update info from instrs

These were going to be used to trigger instrument firmware updates, but
we're going to base all that on subsystems now so they can be removed.

sfoster1 committed:
refactor(hardware): remove pipette lookup fallback

It's now been long enough since we got updated bootloaders that we don't
need the fallback for pipettes with bootloaders from EVT; let's remove
the fallback for looking up the pipette type in the bootloader.

Note: This commit does not stand alone because the callsites for it
aren't modified in the commit (because they're very shortly to be gone).
It's set apart so it stands out as a thing that changed.

sfoster1 committed:
refactor(hardware): put fw update in an object

This looks (and is) extremely awkward, but it lets us decoy this object
instead of having to hang mocks off weird strings, and in that context
it's smooth as silk.

sfoster1 committed:
refactor(hardware): network exposes ok status

We're going to do all this subsystem management, but a subsystem can
appear whether the device is in its application or its bootloader. We
need to expose that information somewhere, so let's start at the bottom
and use the network submodule from opentrons hardware to detect whether
the device is OK or not, and put that information in the device
information it exposes.

sfoster1 committed:
refactor(hardware): split tool detect and resolve

We have this ToolDetector thing. It listens for messages from the head
and returns the kinds of pipettes and gripper that are attached, polling
those devices for their information.

Let's refactor that to split those two things into a detect (which is
fast, and involves no further communication) and a resolve (which
involves lots more further communication). This lets us have a nice
async iterator for the detection and a lovely resolution mechanism that
can be called for each detect event, and then use all that stuff
precisely once for the oneshot.

sfoster1 committed:
refactor(hardware-control): add subsystem manager

If we want to hook a bunch of stuff off subsystems, we need a place to
handle them. We want to control the lifecycle of subsystems and their
updates, and that means we should get a lovely object to do this in.

This object has three jobs:
1. Manage the discovery of the "normal" subsystems on the network (i.e.
not stuff that's normally appearing and disappearing like tools), which
is what the ot3controller.probe_network() method used to do; this will
now use the opentrons_hardware.hardware_control.network
2. Manage those subsystems that do appear and disappear, and get the
additional information that they have. This uses the tools detector
infrastructure that we recently changed, triggered off the outcome of
network probes, and exposes and fuses all that information and turns it
into subsystem data.
3. Update those subsystems, including managing multiple ongoing updates
from multiple callers and handling the detection lifecycles of
subsystems as they get updated.

The primary interface here are the properties, because a key feature is
that this manages lifecycles _on its own_. You don't have to ask to
detect tools, because it's triggered off the head microcontroller
notifications; you don't have to re-refresh after an update, it will do
that.

There are some fun implementation details in here; for instance we use
the ever-exotic condition variable to allow for a bit of controlled
state that represents both whether we've completed a tool detection
cycle and whether the worker task got an error.

This commit doesn't stand alone because of the changes in the types
file; a followup will change how the ot3controller and hardware api work.

sfoster1 committed:
refactor(api): use subsystems as primary objects

Now that we have a subsystem manager, we can use it by turning the
hardware api into considering _subsystems_ as a new first-class object
alongside instruments.

subsystems are what get updated, not instruments or mounts; all those
subsystems being ready to run are what defines the api being ready to
operate, not several separate checks. There's not a lot of new
functionality here, mostly things get removed because they have
been added in new submodules.

Because we now rely on submodules as a first class object and subsystems
can be instruments, we can turn the instrument detection mechanism into
something that queries what subsystems are currently present - a flow
which does not require any canbus traffic and will always be correct
because tools are updated asynchronously from the head.

sfoster1 committed:
refactor(robot-server): add firmware updater

The firmware update manager will be the primary functionality driver of
the new subsystems endpoints. It handles the following needs:

- Clients should be able to see what needs updating
- Clients should be able to update things that need updating
- Clients should be able to make decisions based on updates that were
  started by other clients

The way to do this is by having a two-tier system of process management.
Processes can be either ongoing or completed. While processes are
ongoing, you can access them by the subsystem that they're updating.
Because hardware will raise exceptions if you try and start an update
process for a subsystem that's already being updated, there can only
ever be one ongoing update per subsystem.

When an update completes, whether it succeeds or fails, it is no longer
ongoing and goes over to be stored for the long term (until reboot) by
the id it was created with, and can later be retrieved by that ID.

The process flow for a client would be something like:
1. See that a subsystem needs updating
2. Begin an update
   2a. Either the update is in process, in which case we can get its
   status by subsystem, including its ID, or
   2b. The update begins and we can get its status and ID by subsystem
3. Watch the update by its ID, which will always be valid

Because 3 is by-id, you can't miss an event like the update ending while
you had your laptop closed or something. Because you can access an
update process initially by subsystem, you can display the status of an
update process that was started by someone else.

There's some fun implementation details here, like the use of handles to
make sure that callers can only interact with the ongoing process by one
end of a queue. The idea is anything outside this module either
interacts with a manager singleton or with process handles it gets from
the monitor singleton.

sfoster1 committed:
feat(robot-server): add subsystem management

The new subsystem management endpoints allow clients to see the status,
start the firmware updates, and monitor past and present firmware
updates for the microcontrollers attached to the robot system.

The status routes are at GET /subsystems/status and GET /subsystem/status/:subsystem.
These can be used to see what subsystems are present; and whether each
subsystem is OK. While the acceptable values for GET
/subsystem/:subsystems and the subsystem names returned from GET
/subsystem are technically controlled, it's probably a better idea for
clients to get the names from GET /subsystems/status for later use.

The update monitoring routes are at GET /subsystems/updates/current, GET
/subsystems/updates/current/:subsystem, GET /subsystems/updates/all and
GET /subsystems/updates/all/:id. The difference between current and all
is that curreent is keyed by subsystem and only contains resources for
updates that are running at this precise instant; all contains all
updates since the robot booted, ongoing or done.

Updates can be started at POST /subsystems/updates/:subsystem. Starting
an update here has one of three major outcomes:
1. Everything is fine and the update is started; you get the ID for an
update process, and you get a 201 that points you to the ongoing update
2. There's already an update in progress for this subsystem; you get a
303 that points you to the update for that subsystem
3. The subsystem doesn't need an update, or the subsystem isn't present,
or something like this and you get an error and no subsystem.

If you get an update, you can now monitor its status. Clients should
use either "trying to start an update" or getting an ongoing update for
discovery; for monitoring the update, they should use the update ID.
This is because when an update completes it is no longer ongoing and
therefore no longer accessible via the current updates, so if you miss
the completion it will just disappear. Via ID, it will stick around and
you get a complete status.

Laura-Danielle commented on robot-server/robot_server/errors/global_errors.py:
```diff
@@ -23,3 +23,10 @@ class InvalidRequest(ErrorDetails):
+
+
+class IDNotFound(ErrorDetails):
+    """An error returned when an ID is specified incorrectly."""
```
Is this for command ids?

sfoster1 replied:
right now, just the ids in the updates

SyntaxColoring replied:
Do you have opinions for whether the future robot server should consolidate on this in general, for other kinds of IDs?

We have endpoints like `GET /runs/{runID}/commands/{commandID}`, and if you provide a wrong ID, the error should indicate *which one* was wrong.

As a bonus, it's nice if that indication is machine-readable. To do that, I think we always need to use distinct strings like `RunNotFound` and `CommandNotFound` instead of a catch-all like `IDNotFound`.

Or we could have one `IDNotFound` error everything, and use the `title` or `details` to indicate which ID was wrong, not bothering with machine-readability. 🤷

Laura-Danielle commented on api/src/opentrons/hardware_control/types.py:
```diff
@@ -281,7 +281,7 @@ def set_in_point(self, point: top_types.Point, position: float) -> top_types.Poi
 
 
-class OT3SubSystem(enum.Enum):
+class SubSystem(enum.Enum):
```
Why the name change if this is still strictly Flex related stuff?

sfoster1 replied:
well it doesn't always have to be, and I'd love to get it to not be, and this lets us get out ahead

Laura-Danielle commented on robot-server/robot_server/subsystems/models.py:
```diff
@@ -0,0 +1,138 @@
+)
+
+
+class SubSystem(enum.Enum):
```
Why do we need to re-define subsystem here? Some weird pydantic/fast api issue?

sfoster1 replied:
to make other stuff able to import it without problems, and have this one have string values instead of auto assigned

Laura-Danielle commented on robot-server/robot_server/subsystems/models.py:
```diff
@@ -0,0 +1,138 @@
+        ...,
+        description="What firmware version a prospective update would leave the subsystem running.",
+    )
+    fw_update_needed: bool = Field(
```
Do we need both `ok` and `fw_update_needed` if `ok` signifies that it's out of date?

sfoster1 replied:
i think it's worth being able to tell the two cases apart, and if we're worried about the redundancy i'd almost rather pull the "needs an update" part out of `ok`

Laura-Danielle commented on api/src/opentrons/hardware_control/dev_types.py:
```diff
@@ -64,9 +64,6 @@ class AttachedGripper(TypedDict):
 
 class InstrumentDict(TypedDict):
     display_name: str
-    fw_update_required: bool
```
nioce

Laura-Danielle commented on api/src/opentrons/hardware_control/backends/subsystem_manager.py:
```diff
@@ -0,0 +1,374 @@
+    _tool_task_state: Union[bool, Exception]
+    _updates_required: Dict[FirmwareTarget, FirmwareUpdateRequirements]
+    _updates_ongoing: Dict[SubSystem, UpdateStatus]
+    _update_bag: FirmwareUpdate
```
👜

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -2122,3 +2054,7 @@ def _axis_map_from_ot3axis_map(
         return ret
+
+    @property
+    def attached_subsystems(self) -> Dict[SubSystem, SubSystemState]:
```
Is this required still?

sfoster1 replied:
yeah, this is the way to get the appropriate subsystems out of the api. guess i should probably remember to actually implement it huh

Laura-Danielle reviewed (approved):
The tool detection refactor and firmware updates looks super clean! Had a few comments questions but this looks good to me after a bit of testing.

sfoster1 committed:
actually implement hardware control subsystems

sfoster1 committed:
refactor(api,robot-server): Update fw in server

When we start the robot, we need to run any required firmware updates -
maybe we just got a system update, maybe somebody pushed something,
maybe the system is booting with a new instrument - and we want those
updates to be visible via HTTP. That means they need to be update
processes, and that means that they need to happen via the firmware
update manager in the server.

So let's hoist the initial update out of the hardware controller build
function and up into the robot server system startup tasks, which
requires refactoring them a bit to handle the new ot3-only behavior.

As a side-effect of this refactor, we should no longer be trying to
blink a nonexistent button on the OT-3, and we should have the server
become ready for interaction _much_ faster since it no longer waits for
firmware updates to be complete before sending its systemd ready notice.
While this might cause a weird time in the ODD interaction while
firmware updates are ongoing but the app seems fine, the solution to
that is the frontend changes to view and handle updates.

vegano1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -345,11 +335,8 @@ def _info_for_required_updates(
-# field, both in application and bootloader, we can remove the attached_pipettes from this codepath.
 def check_firmware_updates(
     device_info: Dict[FirmwareTarget, DeviceInfoCache],
-    attached_pipettes: Dict[NodeId, PipetteType],
```
sweet

ahiuchingau replied:
🙌 🙌 🙌

vegano1 commented on hardware/opentrons_hardware/hardware_control/tools/detector.py:
```diff
@@ -46,6 +46,137 @@ async def _await_one_result(callback: WaitableCallback) -> ToolDetectionResult:
+                )
+                seen.add(node)
+                break
+            elif isinstance(response, message_definitions.ErrorMessage):
```
Wouldn't this keep us in an infinite loop if the response is an ErrorMessage

sfoster1 replied:
It would, so this function should always be called with a timeout around it.

vegano1 commented on hardware/opentrons_hardware/hardware_control/tools/detector.py:
```diff
@@ -57,17 +188,31 @@ def __init__(self, messenger: CanMessenger) -> None:
-            await self._messenger.send(
-                node_id=NodeId.head, message=attached_tool_request_message
+            await self.check_once()
+            while True:
```
very cool, will continue to detect instruments as long as the calling function continues to consume this iterator.

vegano1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -1008,53 +893,13 @@ def home_position() -> OT3AxisMap[float]:
-        core_present = set(await self._network_info.probe(expected, timeout))
-        self._present_devices = core_present
-        log.info(f"The present devices are now {self._present_devices}")
+        await self._subsystem_manager.refresh()
```
do we need to continue to explicitly check for new nodes? arent they asynchronously detected now?

sfoster1 replied:
new instruments are asynchronously detected based on the levels of the detection lines changing. but those levels can only be relied upon to change when instruments are added or removed, not necessarily when they reboot or go between bootloader and application. so we definitely can't rely on the asynchronous detection here.

vegano1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -645,7 +578,6 @@ async def _cache_instruments(
             else:
                 self._pipette_handler.hardware_instruments[pipette_mount] = None
 
-        await self._backend.probe_network()
```
nevermind, we are no longer polling !
this is awesome 💯

vegano1 reviewed (approved):
This is incredible work, 
- The async tool detection and caching are sweet
- The integration across the stack makes sense

sfoster1 committed:
refactor(robot-server): get_ot3_hardware fixups

FastAPI actually evaluates the type hints on dependency functions at
runtime. This is a Really Big Problem for us because we can't rely on
OT3API being importable. It means that get_ot3_hardware can't be a
dependency; that also means it can't use dependencies and callers have
to pass in actual objects.

Also, you can't really do isinstance(some_thread_manager, OT3API)
because it's a thread manager not an OT3API; add a passthrough wrapper
for the thread manager that lets you ask what it's wrapping.

sfoster1 committed:
refactor(robot-server): include subsystems router

sfoster1 committed:
refactor(api): can't cache before firmware update

ahiuchingau commented on robot-server/robot_server/instruments/instrument_models.py:
```diff
@@ -34,8 +34,11 @@ class _GenericInstrument(GenericModel, Generic[InstrumentModelT, InstrumentDataT
     serialNumber: str = Field(..., description="Instrument hardware serial number.")
+    subsystem: Optional[SubSystem] = Field(
+        None,
+        description="The subsystem corresponding to this pipette.",
```
```suggestion
        description="The subsystem corresponding to this instrument.",
```

sfoster1 committed:
fix: rebase fixups

SyntaxColoring commented on robot-server/robot_server/subsystems/firmware_update_manager.py:
```diff
@@ -0,0 +1,416 @@
+            )
+        # mypy continues to think this is a method but it is a function stored in an
+        # attribute and therefore does not need a self argument
+        await self._complete_callback()  # type: ignore[misc]
```
Does this need to be in a `finally` block so it runs if the error is a `BaseException`, and not an `Exception`? For example, `KeyboardInterrupt`, or `asyncio.CancelledError` in Python 3.8+.

sfoster1 committed:
feat(robot-server): subsystems for bad instrs

The whole point of the subsystems interface is to have something to
present when an instrument is physically present but not actually able
to operate as an instrument for whatever reason. Since the need for the
subsystems is based on the need to handle instruments, and clients want
to be polling instruments, have /instruments return information about
the subsystem if an instrument needs an update or isn't ok so that
clients need to go look at the subsystems.

SyntaxColoring commented on robot-server/robot_server/service/json_api/response.py:
```diff
@@ -80,7 +80,17 @@ class MultiBodyMeta(BaseModel):
     """A response that returns multiple resources."""
 
-    data: List[ResponseDataT] = Field(..., description=DESCRIPTION_DATA)
+    data: Sequence[ResponseDataT] = Field(..., description=DESCRIPTION_DATA)
+    # note: the type of data is the protocol Sequence even though this is an actual
+    # object. this is normally something you shouldn't do - the type should be described -
+    # but it's done here because the type definition of the constructor and the construct()
+    # non-validating classmethod is taken from the type of this member, and there we really
+    # want the arguments to be Sequence so they can accept narrower subtypes. For instance,
+    # if you define a function as returning SimpleMultiBody[Union[A, B]], you should really
+    # be able to do return SimpleMultiBody.construct([A(), A(), A()]) or even
+    # SimpleMultiBody[Union[A, B]].construct([A(), A(), A()]). However, because construct's
+    # params are defined based on the dataclass fields, the only way to get the arguments
+    # to be covariant is to make data the covariant Sequence protocol.
```
If you want to avoid changing this `List` to `Sequence`, my usual approach to this is to widen the type at the call site.

```python
async def _get_attached_instruments_ot2(
    ...
) -> PydanticResponse[SimpleMultiBody[AttachedItem]]:
    # was:
    # response_data = [...]
    
    # becomes:
    response_data: List[AttachedItem] = [...]
    
    return await PydanticResponse.create(
        content=SimpleMultiBody.construct(
            data=response_data,
            meta=MultiBodyMeta(cursor=0, totalLength=len(response_data)),
        ),
        status_code=status.HTTP_200_OK,
    )
```

sfoster1 replied:
Yeah but I hate that too :/ I think sequence ends up working ok and makes the `SimpleMultiBody` type more useful, it just feels kinda gross

sfoster1 committed:
refactor(robot-server): clean up instr responses

The instrument responses in the bad vs good case were wildly different
which makes for a bad experience with clients. They can now be
distinguished via the value of the `ok` response element - bad is bad,
good is good. "OK" here means "fully ready to operate", aka subsystem.ok
&& (!subsystem.update_required); if clients need to figure out which is
the problem, ask the subsystem.

sfoster1 committed:
test script for running updates

SyntaxColoring commented on robot-server/robot_server/subsystems/router.py:
```diff
@@ -0,0 +1,358 @@
+
+@subsystems_router.get(
+    "/subsystems/updates/all",
+    summary="Get a list of all updates by id.",
+    description="Get a list of all updates, including both current and concluded ones, by their update id. This is a list of all previous updates since the machine booted, including their final status and whether they succeeded or failed. While an update might complete while you're not polling and therefore disappear from /subsystems/updates/running, you can always pull a previous update by id fr ...
```
A couple docstring suggestions:

"Get a list of all updates by ID" means, to me, it will return an object whose keys are the update IDs. But it looks like this is actually returning a list (which is good).

Also, I don't see a `/running` endpoint, but I think it refers to `/current`.

```suggestion
    summary="Get a list of all updates",
    description="Get a list of all updates, including both current and concluded ones, since the machine booted. This includes their final status and whether they succeeded or failed. While an update might complete while you're not polling and therefore disappear from /subsystems/updates/current, you can always pull a previous update by id from here.",
```

sfoster1 replied:
re: the ids - I had not considered the routes with multiple ids. i think maybe subclassing is a good solution there? i think we want to keep machine readability, but i will also say we could put "which thing is wrong" somewhere other than the title of the exception in a machine-readable way.

SyntaxColoring commented on robot-server/robot_server/subsystems/models.py:
```diff
@@ -0,0 +1,138 @@
+class UpdateState(enum.Enum):
+    """The high-level state of an ongoing update."""
+
+    queued = "queued"
```
No opinion, just wondering: what's the client-facing difference between an update that's `"queued"` versus an update that's `"updating"` and has `"progress": 0`?

sfoster1 replied:
Nothing, really, although `progress: 0` is something that shouldn't generally occur.

SyntaxColoring commented on robot-server/robot_server/subsystems/router.py:
I'm finding the URL path hierarchy of these new endpoints a bit confusing.

As written, we have these, which make sense to me:

* `GET /subsystems/status`
* `GET /subsystems/status/{subsystem}`

But then these:

* `POST /subsystems/updates/{subsystem}`
* `GET /subsystems/updates/current`
* `GET /subsystems/updates/current/{subsystem}`
* `GET /subsystems/updates/all`
* `GET /subsystems/updates/all/{id}`

Specifically, I think it's confusing that you `POST` to `/subsystems/updates/{subsystem}` while that path doesn't appear in any `GET`. And also, that the list of current updates is only filterable by subsystem, whereas the list of all updates is only filterable by ID.

I think this is a good use case for optional query parameters. What do you think of this?

* `POST /subsystems/updates`, with a *request body* that designates which subsystem
    * This is analogous to `POST /runs`, which takes the protocol ID via the request body.
* `GET /subsystems/updates?current={true|false}&subsystem={subsystem}&id={id}`
    * `current`, `subsystem`, and `id` may all be omitted, and if they are, the default is to return everything.

sfoster1 replied:
I understand and agree with the `POST` to `/subsystems/updates/{subsystem}` thing, but I strongly disagree with the query params for the get to get updates. The list of current updates is very purposefully filterable only by subsystem, because each subsystem can have at most one ongoing update. The current state of the system is "which subsystems are currently being updated". 

I could agree with the `/all/` part also being filterable by subsystem if you want, but I don't really think of these as filters so much as specifiers. Because you can have at most one update per subsystem, `/current/{subsystem}` is fully specific: it will return zero or one responses. Because ids are unique per run, `/all/{id}` is fully specific: it will return zero or one responses. You get to pick either seeing everything (via `/updates/all`) or one thing (via `/updates/all/{id}`); you get more information for the one-thing, but you have to handle the part where it might not exist.

Having a structural difference between /current and /all also guides clients to the correct behavior, or what I think is the correct behavior: you might glance at the current updates to get a snapshot of what's going on, but in general you want the IDs, and you want to know about specific updates - because you already got the list of updates you care about from `/current`.

What I'd propose is to change the start-an-update route to `POST /subsystems/updates/{current}`. That would also remove the need for a redirect response (it would just be a `200 OK` instead).

sfoster1 committed:
a classic type erasure problem

SyntaxColoring reviewed (commented):
Makin' my way through the `robot-server` diff.

sfoster1 committed:
you have to actually return from loops

sfoster1 committed:
refactor(api): update at most passed arguments

We would unconditionally update anything that needed updating, which is
not what we want; instead, we want to update the intersection of what
needs updating and what was requested to be updated.

sfoster1 committed:
refactor(robot-server): split hw postinit

By putting postinit in a separate task, we can let the server come up
ready earlier. This lets us avoid issues with timeouts and allow clients
to see any details of the postinit tasks that we might want to expose -
for instance, the status of ongoing first-boot firmware updates for
connected subsystems.

sfoster1 committed:
lights need updates

sfoster1 committed:
Add a global exception handler for updates

sfoster1 committed:
hardware: add a consistent comms exc for drivers

This will make catching them easier.

sfoster1 committed:
hardware: handle USB comm errors

If we try and probe while the CANbus has an update going on, that's
fine, those devices just don't respond. But if we try and probe the USB
while the rear panel is updating, we can't find the usb device and will
get an exception. Since updates do a network probe when they complete,
if we have multiple concurrent updates then if the USB update doesn't
complete first the others will try and fail to probe it.

But really, all a communication error via USB means is that the probe
failed - that device is not present. So let's treat it like that.

sfoster1 committed:
refactor(hardware,api): probe_specific

When we probe the network, we send a broadcast ping and wait for
anything. We're askign the network in general what's present.

But with the subsystem manager, in a lot of situations what we really
want is to ask whether _specific devices_ are present, since we're
maintaining an estimate of what's present before then. We only want to
do the full broadcast when we start.

So add a new method called probe_specific, which instead of broadcasting
and returning _anything_ (and yelling at you if it misses what you
expect) will broadcast and look _only for responses from what you tell
it_, and not touch the rest of cached state.

This can then be used when a firmware update completes to prevent the
pings from looking for other devices in the middle of their own updates.

sfoster1 committed:
robot-server format

sfoster1 committed:
refactor: check devices more leniently after update

When we update, we check that all the devices that should have just
completed an update have come back up and are present on the network.
This lets us verify that the update succeeded. But they're rebooting
after an update, so we need to give them a little longer to start with a
nice retry system.

Also, add a mechanism to indicate that specific tools are no longer present.

sfoster1 committed:
refactor: exclusive can access for restarts

There's a period when our microcontroller applications and bootloaders
start where their can peripherals are running and accepting messages,
but no filters have been loaded. This usually isn't a problem, but it's
a really big problem when it's the bootloader, and at the same time
there's an update running and sending data for a different node - in
that case, the starting device gets a message for the other one and
reads it in. This doesn't cause corruption because we're checking
checksums and total packet counts, but it does make the update fail.

To fix that, we can introduce the concept of exclusive access to the can
messenger, since can messengers are shared. This pretty much boils down
to a lock. send() and ensure_send() take the lock all the time; but also
you can take the lock explicitly and prevent use of the canbus for some
longer time, and then use exclusive_send and exclusive_ensure_send to
avoid deadlocks.

By holding exclusive access in the updater while we're telling devices
to enter the bootloader, we should prevent any udpate data packets from
hitting the bus in the vulnerable period while a device is restarting.

There's some improvements that can be made here - there's no reason to
prevent concurrent access to the canbus when nothing's holding the
exclusive lock - but this works for now.

smb2268 reviewed (approved):
Updates to the instrument model look good to me!

sfoster1 committed:
refactor: update progress fix

The progress needs to be intified and multiplied in hardware control,
since that's where we turn things into ints.

sfoster1 committed:
refactor: timeout tweak

It's now pretty important to be on the nose about noticing something has
rebooted since we have exclusive access so make shorter timeouts and
more retries

sfoster1 committed:
really need a sleep here so we don't spam

sfoster1 committed:
handle BaseExceptions in updates

sfoster1 committed:
fix a missing f for the fstring

sfoster1 committed:
fix up the lovely script

sfoster1 committed:
some little docstring changes

sfoster1 committed:
handle ot2-only tests

sfoster1 committed:
ot3 only test for now

sfoster1 committed:
bump a timeout

sfoster1 committed:
make some tests ot3-only

sfoster1 committed:
Revert "bump a timeout"

This reverts commit 53b8a9d3448aafd00f556f0c8868a4b6b23e15b6.
